### DISCUSSION TITLE
Linearize displaced mesh with MultiApp applications

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#18416] [Q&A Meshing] [answered]
Hello,

I have a situation where im using MultiApps to transfer variables from an app that uses a displaced mesh to model a flow loop to an app i am developing which looks to linearize the mesh and use periodic boundary conditions to simulate the "loop" part. Bellow i have a simple figure which only show part of the loop, but visualizes the issue. the black circles represent the mesh nodes. 

![mesh](https://user-images.githubusercontent.com/36215108/126816890-a7c0ef22-87a1-47ae-8ddc-1856ace8f54f.png)

I am having trouble trying to figure out how to transfer variables from the main app to the sub app. I assume `MultiAppNearestNodeTransfer` finds the nearest node by x, y, z coordinates and im not sure how to do this with positions in `MultiAppMeshFunctionTransfer`. In the figure each app has the same number of mesh nodes however, in my case i would like to use more nodes in my sub app. As i mentioned before, the main app uses a displaced mesh, but my sub app would not. I would also like the flexibility to have my app be the main app and the other app to be the sub app. Can anyone give me some guidance in solving this problem?

Thanks,
Zack

**Comment by GiudGiud** [comment]
Hi Zack

Is this rotation the only transformation you care about or is there going to be arbitrary transforms in space?
Do the initial meshes match exactly at the beginning of the simulation ?

Guillaume

**Reply by rtaylo45** [reply]
In the picture I am only showing the elbow example but in reality the main apps mesh is a loop. So it has another elbow which completes the loop.   So to answer your questions it would be arbitrary transforms in space. I would need to conserve the total length of the loop.  They utilize a displaced mesh and connect the variables within each sub mesh using boundary conditions and constraints, to simulate a loop. 

Both meshes would be static and would not match exactly at the beginning of the simulations. The main app would have mesh coordinates which visually represent a loop. The submesh would only have a single dimension, with periodic BCs at each end. 

On other note is that the main apps mesh condaints sub meshes which could include solid areas for wall conduction and fluid areas for flow. My subapp is only modeling physics which are in the fluid domain. 

Zack

**Reply by GiudGiud** [reply]
This all seems difficult for mesh to mesh transfer. Like you would need to code your own transfer. 

An approximate transfer could use vectorpostprocessors, which are easier to arbitrarily distribute in space then match one by one. I ll let others comment on potential solutions. 

If you could make the subapp mesh match instead, it would be much easier. Are you doing this 1D periodic subapp mesh to lower computational cost?

**Reply by rtaylo45** [reply]
Ok. If the main app uses a displaced mesh, how would i go about generating their mesh for my sub app to use? I tried using `--mesh-only` but it seems the the displaced information is not being saved in the `in.e` file. Because the mesh is displaced could i used Dirichlet boundary conditions on the nonlinear variables in my sub app to "connect" the mesh displacements?

**Reply by GiudGiud** [reply]
If you can write the mesh in a sub app as the transform of the mesh in the main app, you could use a **modified** https://mooseframework.inl.gov/source/meshgenerators/TransformGenerator.html , which implements the transform of your choice. Then you could do the transfer on the same mesh. 

I don't know about the displaced mesh way. I asked the rest of the team for input.

**Reply by GiudGiud** [reply]
Did you set use_displaced in the exodus output block?

For the transfer problem, I've reached out and the recommendation is to a NearestPointLayeredAverage uo with a MultiAppUserObjectTransfer. 
https://mooseframework.inl.gov/source/userobject/NearestPointLayeredAverage.html

This will allow to match points between the two different meshes

**Reply by rtaylo45** [reply]
I did set `use_displaced` in the exodus block and when i run the input file with `--mesh-only` and view the `in.e` file, the mesh is not the same. I attached the generated mesh using `--mesh-only` and the main apps solution if i were to run the input file. 

[files.zip](https://github.com/idaholab/moose/files/6894449/files.zip)

**Comment by aeslaughter** [comment]
Are you doing this just to get the periodic boundary condition? If so, it might be better to work on creating the correct Periodic Boundary condition. If this is the case, we should be able to connect the boundaries in the elbow.

If you indeed need to do the transfer, based on your figure it is a one-to-one transfer, so the MultiAppCopyTransfer would just work. But, I am also unsure of how to handle the displacements correctly. @fdkong do you know how to transfer displacements? Is it possible to create a Transfer, using the MultiAppFieldTransfer with the displaced problem supplied to the transfer method to get the desired result.

**Reply by rtaylo45** [reply]
@aeslaughter @GiudGiud So I kinda look at solving this problem in two ways. 

1. Use the main apps displaced mesh with one-to-one transfer so that all of the nodal variables line up so that i can do variable transfer easy. I could possibly use already developed moose boundary conditions such as `ADMatchedValueBC` to "connect" the variables on the displaced mesh boundaries?
2. Generate a mesh that is used for my subapp that is different than the main apps mesh. Then i would need to make sure that i can properly transfer the variables from the main apps mesh to my subapp. 

I was suggesting using periodic boundary conditions at the bottom and top boundaries on the linear subapps mesh to simulate the loop effect of the main apps problem. Im still new to moose, so im not sure of the best way to solve this problem and not sure what transfer tools i can use to do this. In any case, im not sure which method, either 1 or 2, would be the best way to solve this problem. If i go with option 1, would i need to generate a mesh file for my subapp to read? or could I simply just use MultiApps to transfer the main apps mesh to the sub app and could I do either of those with a displaced mesh? If i go option 2, i feel like it would be a lot of work to transfer the variables from the main apps mesh onto the subsapps generated mesh. I would probably need to write my own transfer class? It looks like i might be able to do what @GiudGiud said and use NearestPointLayeredAverage to transfer the variables from the main app to the subapp when they have different meshes. Is this true?

**Reply by fdkong** [reply]
OK, there are two questions here (not one question)

1. Periodic boundary conditions in subapp.  We already have this capability. You could use something like this 

```

[BCs]
  [./Periodic]
    [./all]
      variable = u 
      auto_direction = 'x'
    [../]
  [../]
[]
```

2. Transfer between main-app and subapp. You might try to make the mesh nodes match up first and then use MultiAppCopyTransfer. Once this works, we can then consider other transfer options. There are plenty of transfers.

**Reply by rtaylo45** [reply]
Yes i am aware of periodic boundary conditions in moose/subapps. My question is not about using periodic boundary conditions. Part of my question is how to transfer the displaced mesh from the main app to the sub app. I tried using the `--mesh-only` flag when running the executable but it does not seem to properly create a mesh that i can use. I thing this is because the main app is using a displaced mesh.

**Reply by GiudGiud** [reply]
mesh only cuts the simulation very early, possibly before the transfers occur. I ll ask for transferring a displaced mesh

**Reply by GiudGiud** [reply]
Ok could you try using the `clone_master_mesh` parameter of the multiapp? 
This should allow you to use the displaced mesh in the subapp

**Reply by rtaylo45** [reply]
So this is how the the multiapp block is set up in my main app. 
```
[MultiApps]
  [mole]
    type = FullSolveMultiApp
    app_type = moleApp
    input_files = 'mole_sub.i'
    execute_on = 'FINAL'
    use_displaced_mesh = true
    clone_master_mesh = true
  []
[]
```
As it still is not transfer the displaced mesh. This is the example output.
![Screen Shot 2021-08-04 at 4 58 40 PM](https://user-images.githubusercontent.com/36215108/128253944-f59bd84c-13fc-4f57-8743-da7e43be0b3c.png)

**Reply by GiudGiud** [reply]
@fdkong do you know about displaced mesh transfer?

**Comment by travismui** [comment]
@rtaylo45 In case someone else from the SAM team hasn't dropped you a message, we've added an `--output_displaced_mesh` runtime flag in SAM to output the displaced mesh and quit, similar to the expected behavior from `--mesh-only`. This update has been pushed to SAM's `devel` branch so hopefully this helps with this specific issue, let us know if you run into any problems.

**Reply by GiudGiud** [reply]
Hi Travis, this seems like something that would be helpful for every application all across the framework if you could consider pushing it upstream?

**Reply by travismui** [reply]
Hi Guillaume– I did not work on this update in SAM, but from what I can tell our modification was to add an action (within SAM) that points to the displaced mesh for the mesh output function. Taking a look at the framework, I believe the only real difference lies at [this line](https://github.com/idaholab/moose/blob/0e5f0880f25b7c7512874866fb9a1e9c63f2eea2/framework/src/actions/MeshOnlyAction.C#L36) in the `MeshOnlyAction`, for which our action points to ` _action_warehouse.displacedMesh()` rather than `  _action_warehouse.mesh()`.

**Comment by GiudGiud** [comment, marked as answer]
@rtaylo45 none of what we proposed worked out right?

@fdkong is going be working on a coordinate mapping system in FY22 so that we can map main-app coordinates to any sub app coordinates. This should solve your case and @vincentlaboure.

The relevant issue in MOOSE is https://github.com/idaholab/moose/issues/12293

**Reply by rtaylo45** [reply]
No, none of it did. But we are going a different route to solving the problem. We are going to compile the two codes together, this ended up making the most sense because of some other SAM related reasons. When the coordinate mapping system becomes available, it would be nice to go back and revisit this problem. But for now we are going to compile the codes together.  Thanks for your help on this issue.
